lib/icinga/apiactions.cpp

```diff
@@ -571,10 +592,13 @@ Dictionary::Ptr ApiActions::ExecuteCommand(const ConfigObject::Ptr& object,
 		endpoint = HttpUtility::GetLastParameter(params, "endpoint");
 
 	MacroProcessor::ResolverList resolvers;
+	Dictionary::Ptr macros;
 	if (params->Contains("macros")) {
-		Value macros = HttpUtility::GetLastParameter(params, "macros");
-		if (macros.IsObjectType<Dictionary>())
-			resolvers.emplace_back("override", macros);
+		Value vmacros = HttpUtility::GetLastParameter(params, "macros");
+		if (vmacros.IsObjectType<Dictionary>()) {
+			resolvers.emplace_back("override", vmacros);
+			macros = vmacros;
```

Comment:
No, not **those** macros! See [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/icinga/checkable-check.cpp#L555-L584) how to get the macros needed...

Reply:
Hi @Al2Klimov,
sorry but I did not understand exactly what your linked section does, unless `GetCheckCommand()->Execute(this, cr, macros, false);` does something magical with the macros, isn't `macros` just an empty Dictionary?

Maybe it's a misunderstanding, we're a just a little insecure about what you mean with _those_ and _other_ macros and would like to understand what to do and how it should work.

Thanks!

Reply:
Remember the arguments like `"--host" = "$address$"` in command configs?

`theResolvedCommand->Execute(checkable, nullptr, execMacros, false)` resolves all such of `theResolvedCommand` in `checkable` context and stores them into `execMacros` instead of actually running `theResolvedCommand`.

`execMacros` is what you need to send to the agent via "macros".

The macros are resolved [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/methods/pluginchecktask.cpp#L31-L45) and [here](https://github.com/Icinga/icinga2/blob/fd4abff369c7268d1c61219713b73c3c02c4b22f/lib/icinga/pluginutility.cpp#L27-L28), so for overriding you'll need to add `overrideMacros` parameters (Dictionary::Ptr, default nullptr) to all functions across the call stack, from `Execute()` (what you're calling directly) to `Plugin*Task::ScriptFunc()` (which sets up the resolver).

Reply:
Not to add too much complexity you may even add a thread-local variable which is set (and reset via Defer!) by this action and read by `Plugin*Task::ScriptFunc()`.

Reply:
... as static member of MacroResolver.

Reply:
And this one?

Reply:
Whoops, sorry... on it!

Reply:
We didn't get exactly the point, did you mean like this? https://github.com/Icinga/icinga2/pull/8064/commits/be2701bc8acf75f8210b7ae885704337090d723a
Otherwise could you please elaborate a bit more? 
Thank you